### PR TITLE
[1.16] Stop serving deprecated beta workload APIs

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

* Disables default serving of apps/v1beta1, apps/v1beta2, and deprecated extensions/v1beta1 APIs

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

xref https://github.com/kubernetes/kubernetes/issues/43214

**Does this PR introduce a user-facing change?**:
```release-note
The following APIs are no longer served by default:
* All resources under `apps/v1beta1` and `apps/v1beta2` - use `apps/v1` instead
* `daemonsets`, `deployments`, `replicasets` resources under `extensions/v1beta1` - use `apps/v1` instead
* `networkpolicies` resources under `extensions/v1beta1` - use `networking.k8s.io/v1` instead
* `podsecuritypolicies` resources under `extensions/v1beta1` - use `policy/v1beta1` instead

Serving these resources can be temporarily re-enabled using the `--runtime-config` apiserver flag. 
* `apps/v1beta1=true`
* `apps/v1beta2=true`
* `extensions/v1beta1/daemonsets=true,extensions/v1beta1/deployments=true,extensions/v1beta1/replicasets=true,extensions/v1beta1/networkpolicies=true,extensions/v1beta1/podsecuritypolicies=true`

The ability to serve these resources will be completely removed in v1.18.
```

### CODE REVIEW

**Comment by liggitt** [comment]
green integration :)

**Comment by liggitt** [comment]
> pull-kubernetes-e2e-gce — Job succeeded.

🎉

**Comment by shyamjvs** [comment]
Presubmit seems to fail due to:

```
20:21:18.651800   81226 clusterloader.go:233] Error while setting up prometheus stack: error while applying (/go/src/k8s.io/perf-tests/clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml): timed out waiting for the condition
```

Could be related to recent enabling of prometheus in scale tests - https://github.com/kubernetes/test-infra/pull/12394

fyi - @mm4tt @wojtek-t

**Comment by liggitt** [comment]
Yes. The Prometheus manifest uses a deprecated version. Updated to apps/v1 in https://github.com/kubernetes/perf-tests/pull/525

https://github.com/kubernetes/perf-tests/pull/525

**Comment by mm4tt** [comment]
@liggitt, let me know if you need help debugging Prometheus issues in failing scalability tests

**Comment by liggitt** [comment]
@mm4tt if you can dig into it, that would be extremely helpful. If I had to guess, I suspect the prometheus operator is depending on one of the deprecated extensions/v1beta1, apps/v1beta1, or apps/v1beta2 APIs this PR turns off.

**Comment by mm4tt** [comment]
Np, will take a look today.

**Comment by mm4tt** [comment]
The error I see in the prometheus operator logs:
```
E0508 08:14:49.636351       1 reflector.go:134] github.com/coreos/prometheus-operator/pkg/prometheus/operator.go:407: Failed to list *v1beta2.StatefulSet: the server could not find the requested resource
E0508 08:14:49.638320       1 reflector.go:134] github.com/coreos/prometheus-operator/pkg/alertmanager/operator.go:233: Failed to list *v1beta2.StatefulSet: the server could not find the requested resource
```

Looks like this change would break the prometheus-operator, even at master it still uses the apps/v1beta2, e.g. ([link](https://github.com/coreos/prometheus-operator/blob/master/pkg/prometheus/operator.go#L273 ))
`
return cache.NewListWatchFromClient(c.kclient.AppsV1beta2().RESTClient(), "statefulsets", namespace, fields.Everything())
`

@liggitt, what's the plan for updating external api clients? (like prometheus-operator)?

**Comment by liggitt** [comment]
> Looks like this change would break the prometheus-operator, even at master it still uses the apps/v1beta2, e.g. ([link](https://github.com/coreos/prometheus-operator/blob/master/pkg/prometheus/operator.go#L273))
> `return cache.NewListWatchFromClient(c.kclient.AppsV1beta2().RESTClient(), "statefulsets", namespace, fields.Everything())`

Yeah, that has to be fixed (all apps/v1beta1, apps/v1beta2 resources, and extensions/v1beta1 resources other than Ingress are *long*-deprecated and will be turned off in this PR). Since a hard dependency on that operator got added to our CI, we need to ensure it gets fixed asap and update to the new release.

> @liggitt, what's the plan for updating external api clients? (like prometheus-operator)?

The two-year deprecation period gives external projects time to update. As we're aware of cases like this, we can file issues to make them aware.

If they are projects we are using ourselves, we have to make sure they get fixed.

**Comment by liggitt** [comment]
opened https://github.com/coreos/prometheus-operator/issues/2592. @mm4tt, can you drive updating the operator to non-deprecated versions?

**Comment by mm4tt** [comment]
> opened [coreos/prometheus-operator#2592](https://github.com/coreos/prometheus-operator/issues/2592). Matt Matejczyk, can you drive updating the operator to non-deprecated versions?

Sure, will take care of it

**Comment by mm4tt** [comment]
https://github.com/coreos/prometheus-operator/pull/2593 has been merged today and there is a big chance that it will get released next week. Once it happens it should unblock us.

**Comment by liggitt** [comment]
@mm4tt can you let me know when the operator is updated? thanks.

**Comment by mm4tt** [comment]
@liggitt, https://github.com/kubernetes/perf-tests/pull/533 merged yesterday evening, but looks like you've already noticed :) Happy to see it works!

**Comment by liggitt** [comment]
Getting this in early in the 1.16 timeframe, adding reviewers.

/assign @kow3ns 
for workloads APIs

/assign @thockin 
for network policy

/assign @tallclair 
for pod security policy

**Comment by tallclair** [comment]
LGTM for PSP

**Comment by smarterclayton** [comment]
/approve

I am excited to see these go.

File: test/integration/master/synthetic_master_test.go [hunk 1]
```diff
@@ -523,15 +523,9 @@ func TestAppsGroupBackwardCompatibility(t *testing.T) {
 		expectedStatusCodes map[int]bool
 		expectedVersion     string
 	}{
-		// Post to extensions endpoint and get back from both: extensions and apps
-		{"POST", extensionsPath("deployments", metav1.NamespaceDefault, ""), deploymentExtensions, integration.Code201, ""},
-		{"GET", extensionsPath("deployments", metav1.NamespaceDefault, "test-deployment1"), "", integration.Code200, "extensions/v1beta1"},
-		{"GET", appsPath("deployments", metav1.NamespaceDefault, "test-deployment1"), "", integration.Code200, "apps/v1"},
-		{"DELETE", extensionsPath("deployments", metav1.NamespaceDefault, "test-deployment1"), "", integration.Code200, "extensions/v1beta1"},
 		// Post to apps endpoint and get back from both: apps and extensions
```
Comment (janetkuo) [inline test/integration/master/synthetic_master_test.go]: nit: get back from apps

**Review by janetkuo — APPROVED** [review]
One nit, LGTM otherwise

File: test/integration/master/synthetic_master_test.go (hunk 1 above)
Comment (liggitt) [reply test/integration/master/synthetic_master_test.go]: fixed

**Comment by liggitt** [comment]
tagging per:
* https://github.com/kubernetes/kubernetes/pull/70672#issuecomment-501928114
* https://github.com/kubernetes/kubernetes/pull/70672#pullrequestreview-250251711
* https://github.com/kubernetes/kubernetes/pull/70672#issuecomment-501955966

**Comment by neolit123** [comment]
would it make sense to have this as an item under "action required"?

existing manifests (e.g. CNI deployments) trying to be applied against latest k/k master will fail if they are using: `extensions/v1beta1`. this hints that CNI providers may have to adapt.

**Comment by liggitt** [comment]
Definitely, tagged

**Comment by BenTheElder** [comment]
"CNI providers may..." = We goofed that in kind. Whoops 🙃 .. fixed now. Hopefully everyone else long since migrated when the APIs went stable 😉

**Pull request merged**